fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/ScalarFunctionValue.java

```diff
@@ -0,0 +1,332 @@
+        GREATEST_BOOLEAN(ScalarFunction.GREATEST, TypeCode.BOOLEAN, args -> greatest(args, Boolean.class)),
+        GREATEST_STRING(ScalarFunction.GREATEST, TypeCode.STRING, args -> greatest(args, String.class)),
+        GREATEST_FLOAT(ScalarFunction.GREATEST, TypeCode.FLOAT, args -> greatest(args, Float.class)),
+        GREATEST_DOUBLE(ScalarFunction.GREATEST, TypeCode.DOUBLE, args -> greatest(args, Double.class)),
+
+        LEAST_INT(ScalarFunction.LEAST, TypeCode.INT, args -> least(args, Integer.class)),
+        LEAST_LONG(ScalarFunction.LEAST, TypeCode.LONG, args -> least(args, Long.class)),
+        LEAST_BOOLEAN(ScalarFunction.LEAST, TypeCode.BOOLEAN, args -> least(args, Boolean.class)),
+        LEAST_STRING(ScalarFunction.LEAST, TypeCode.STRING, args -> least(args, String.class)),
+        LEAST_FLOAT(ScalarFunction.LEAST, TypeCode.FLOAT, args -> least(args, Float.class)),
+        LEAST_DOUBLE(ScalarFunction.LEAST, TypeCode.DOUBLE, args -> least(args, Double.class)),
+
+        COALESCE_INT(ScalarFunction.COALESCE, TypeCode.INT, args -> coalesce(args, Integer.class)),
+        COALESCE_LONG(ScalarFunction.COALESCE, TypeCode.LONG, args -> coalesce(args, Long.class)),
+        COALESCE_BOOLEAN(ScalarFunction.COALESCE, TypeCode.BOOLEAN, args -> coalesce(args, Boolean.class)),
+        COALESCE_STRING(ScalarFunction.COALESCE, TypeCode.STRING, args -> coalesce(args, String.class)),
+        COALESCE_FLOAT(ScalarFunction.COALESCE, TypeCode.FLOAT, args -> coalesce(args, Float.class)),
+        COALESCE_DOUBLE(ScalarFunction.COALESCE, TypeCode.DOUBLE, args -> coalesce(args, Double.class));
+
+        @Nonnull
+        private final ScalarFunction scalarFunction;
+
+        @Nonnull
+        private final TypeCode type;
+
+        @Nonnull
+        private final Function<List<Object>, Object> evaluateFunction;
+
+        PhysicalOperator(@Nonnull final ScalarFunction scalarFunction,
+                         @Nonnull final TypeCode type,
+                         @Nonnull final Function<List<Object>, Object> evaluateFunction) {
+            this.scalarFunction = scalarFunction;
+            this.type = type;
+            this.evaluateFunction = evaluateFunction;
+        }
+
+        @Nonnull
+        public ScalarFunction getScalarFunction() {
+            return scalarFunction;
+        }
+
+        @Nonnull
+        public TypeCode getResultType() {
+            return type;
+        }
+
+        @Nullable
+        public Object eval(List<Object> args) {
+            return evaluateFunction.apply(args);
+        }
+
+        @SuppressWarnings("unchecked")
+        private static <T extends Comparable<T>> T greatest(final List<Object> args, Class<T> clazz) {
+            T max = (T) args.get(0);
+            for (Object i : args) {
+                if (i == null) {
+                    return null;
+                }
+                if (((T) i).compareTo(max) > 0) {
```

Comment:
I think it does not matter much but the original idea behind physical operators was to not do dispatch on the type but to statically encode what they need to do. All of these physical implementations dynamically dispatch and are effectively the same for all types.

Reply:
Is there a better way to do this without duplicating the code for all types?